Work around malformatted [1] diagramless puzzles.

jpd236 opened the pull request:
If a puzzle has:

-The "diagramless" bit set to true
-No timer information or the timer set to 0
-A user grid with all the black squares perfectly entered by the
"user" but no letters entered

Then it is highly likely to be malformatted, so replace all the
user-entered black squares with blanks as though the user's section
were filled with ':' instead of '.'. This prevents the grid from being
immediately visible upon opening such a puzzle, which spoils the
diagramless aspect entirely.

[1] https://twitter.com/thatpuzzleguy/status/709389211493752832

jpd236 commented:
I realize this is a kind of ugly hack and am open to different ideas or not taking the merge altogether. It's just been frustrating opening these puzzles in XWord without thinking about it and having the grid spoiled, and it seems like this should be a safe way to rectify that.

mrichards42 commented:
I don't have a huge problem with this.  Usually I like handling unusual situations (like bad checksums) without alerting the user, but here I think it would be a good idea, since like you said, this is kind of hacky and only using a heuristic.

I think this code should go in MyFrame::LoadPuzzle instead of load_puz.  Since we (sort of) support multiple file types, it's possible that bad diagramless puzzles could crop up elsewhere, though unless something has changed in the last few years, jpz doesn't support diagramles, and I imagine that almost all of the puzzles users are solving are either jpz or puz.

I have two ideas, but I don't really prefer one over the other
(1) Prompt the user when this situation occurs and ask them what to do.
(2) Go ahead and empty the user grid, but add a menu item that is only available for diagramless puzzles that fills in the black squares.

jpd236 commented:
I'm okay with either as well, but (2) seems interesting as I guess someone who tries and gets frustrated with a diagramless puzzle might appreciate a "make this a diagramed puzzle" button to solve it normally. And that would provide a workaround for any mistakes this might make.

The puzzle author on the twitter thread mentioned he'd look into the other fix, so I may wait to see if he's able to resolve it on his end, but if not, I'll go ahead with that. Thanks for taking a look!

jpd236 commented:
Implemented approach (2). The part about fixing the focused word after the conversion feels iffy to me, so definitely open to a better approach there.

I agree with generalizing to support multiple formats, but think that Puzzle::DoLoad is a better home for this than MyFrame::LoadPuzzle, since it feels like something the puzzle library should handle before the puzzle is returned to the UI layer, so I moved it there. Let me know if you disagree.

mrichards42 commented on puz/Puzzle.cpp:
```diff
@@ -568,4 +569,60 @@ const Puzzle::FileHandlerDesc Puzzle::sm_saveHandlers[] = {
+         square = square->Next())
+    {
+        if ((square->IsSolutionBlack() != square->IsBlack())
+            || (!square->IsBlack() && !square->IsBlank())) {
```
I'm continuing to regret the choice to use 'blank' since it looks so much like 'black'

mrichards42 commented:
Having that code in the puzzle library is fine.  Comments about grid focusing in the code

mrichards42 commented on src/MyFrame.cpp:
```diff
@@ -2025,6 +2038,35 @@ MyFrame::OnRevealLetter(wxCommandEvent & WXUNUSED(evt))
+
+    // Ensure focus moves to a different word as it may be on a black square. We move twice to
+    // ensure that the old word is cleared, as it may not have the correct boundaries.
+    m_XGridCtrl->SetFocusedSquare(m_XGridCtrl->LastWhite(), puz::ACROSS);
```
XGridCtrl::SetFocusedSquare tries to re-use the currently focused word if you don't supply it a new word, and it doesn't check to make sure it's a valid word (since changing from diagramless to normal isn't an expected behavior).  The easiest fix is to supply the word to SetFocusedSquare:

```c++
puz::Square * square = m_XGridCtrl->GetFocusedSquare();
if (square == NULL || square->IsBlack())
    square = m_XGridCtrl->FirstWhite();
m_XGridCtrl->SetFocusedSquare(square, m_puz.FindWord(square));
```

jpd236 replied:
Thanks - was hoping you'd be able to point me to something like this. Done.

mrichards42 commented on src/MyFrame.cpp:
```diff
@@ -2025,6 +2038,35 @@ MyFrame::OnRevealLetter(wxCommandEvent & WXUNUSED(evt))
+    // Send puzzle updated event
+    wxPuzEvent evt(wxEVT_PUZ_LETTER, GetId());
+    evt.SetSquare(m_XGridCtrl->GetFocusedSquare());
+    GetEventHandler()->ProcessEvent(evt);
```
You should be able to simplify this whole bit to m_XGridCtrl->SendEvent(wxEVT_PUZ_LETTER) . . . I'm seeing a similar block of code in MyFrame::OnEraseGrid, which I'm guessing was written before I added that function to XGridCtrl, so that's obviously my fault for not fixing OnEraseGrid

jpd236 replied:
Yep, that felt like the closest analogy to what I was doing here, so I based things off of it. Updated both places.
